Adopt NSTextPlaceholders for Mac.

megangardner opened the pull request:
#### d94443d3d4bf19b813546e7089bb65f4a00b58c3
<pre>
Adopt NSTextPlaceholders for Mac.
<a href="https://bugs.webkit.org/show_bug.cgi?id=272961">https://bugs.webkit.org/show_bug.cgi?id=272961</a>
<a href="https://rdar.apple.com/126696059">rdar://126696059</a>

Reviewed by Wenson Hsieh.

Adopt NSTextPlaceholders which is a parallel of UITextPlaceholders.
Move the now common code to Cocoa to allow for code sharing.

* Source/WebCore/PAL/pal/spi/mac/NSTextInputContextSPI.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView insertTextPlaceholderWithSize:completionHandler:]):
(-[WKWebView removeTextPlaceholder:willInsertText:completionHandler:]):
* Source/WebKit/UIProcess/Cocoa/WKTextPlaceholder.h: Renamed from Source/WebKit/UIProcess/ios/WKTextPlaceholder.h.
* Source/WebKit/UIProcess/Cocoa/WKTextPlaceholder.mm: Renamed from Source/WebKit/UIProcess/ios/WKTextPlaceholder.mm.
(-[WKTextPlaceholder initWithElementContext:]):
(-[WKTextPlaceholder elementContext]):
(-[WKTextPlaceholder rects]):
* Source/WebKit/UIProcess/Cocoa/WKTextSelectionRect.h: Renamed from Source/WebKit/UIProcess/ios/WKTextSelectionRect.h.
* Source/WebKit/UIProcess/Cocoa/WKTextSelectionRect.mm: Renamed from Source/WebKit/UIProcess/ios/WKTextSelectionRect.mm.
(-[WKTextSelectionRectCustomHandleInfo initWithFloatQuad:isHorizontal:]):
(-[WKTextSelectionRectCustomHandleInfo bottomLeft]):
(-[WKTextSelectionRectCustomHandleInfo topLeft]):
(-[WKTextSelectionRectCustomHandleInfo bottomRight]):
(-[WKTextSelectionRectCustomHandleInfo topRight]):
(-[WKTextSelectionRect initWithCGRect:]):
(-[WKTextSelectionRect initWithSelectionGeometry:scaleFactor:]):
(-[WKTextSelectionRect _path]):
(-[WKTextSelectionRect writingDirection]):
(-[WKTextSelectionRect range]):
(-[WKTextSelectionRect _customHandleInfo]):
(-[WKTextSelectionRect rect]):
(-[WKTextSelectionRect containsStart]):
(-[WKTextSelectionRect containsEnd]):
(-[WKTextSelectionRect isVertical]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::insertTextPlaceholder):
(WebKit::WebPageProxy::removeTextPlaceholder):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/AppKitSoftLink.h:
* Source/WebKit/UIProcess/ios/AppKitSoftLink.mm:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::insertTextPlaceholder): Deleted.
(WebKit::WebPageProxy::removeTextPlaceholder): Deleted.
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::insertTextPlaceholderWithSize):
(WebKit::WebViewImpl::removeTextPlaceholder):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::insertTextPlaceholder):
(WebKit::WebPage::removeTextPlaceholder):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::insertTextPlaceholder): Deleted.
(WebKit::WebPage::removeTextPlaceholder): Deleted.
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextPlaceholderTests.mm: Added.

Canonical link: <a href="https://commits.webkit.org/277879@main">https://commits.webkit.org/277879@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/74576adbce24210dec8ee756bb5e75d68bb3e5d8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48815 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28028 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51782 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51501 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44881 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51120 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33971 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25556 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/39904 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49397 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25674 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42093 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21005 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23137 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/43276 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6870 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45068 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43764 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53413 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23866 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20135 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47208 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25129 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42301 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46152 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10758 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25937 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24849 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/PAL/pal/spi/mac/NSTextInputContextSPI.h:
```diff
@@ -29,6 +29,24 @@
+@optional
+- (void)insertTextPlaceholderWithSize:(CGSize)size completionHandler:(void (^)(NSTextPlaceholder *))completionHandler;
+- (void)removeTextPlaceholder:(NSTextPlaceholder *)placeholder willInsertText:(BOOL)willInsertText completionHandler:(void (^)(void))completionHandler;
+@end
```
This should be inside a `#if HAVE(NSTEXTPLACEHOLDER)`.

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/WKTextPlaceholder.mm:
```diff
@@ -26,7 +26,7 @@
 #import "WKTextPlaceholder.h"
 
-#if PLATFORM(IOS_FAMILY)
+#if PLATFORM(IOS_FAMILY) || PLATFORM(MAC)
```
`#if PLATFORM(COCOA)`

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/WKTextPlaceholder.mm:
```diff
@@ -48,11 +48,19 @@ - (instancetype)initWithElementContext:(const WebCore::ElementContext&)context
 {
-    return @[ adoptNS([[WKTextSelectionRect alloc] initWithCGRect:_elementContext.boundingRect]).get() ];
+#if PLATFORM(IOS_FAMILY)
+    return @[(UITextSelectionRect *)adoptNS([[WKTextSelectionRect alloc] initWithCGRect:_elementContext.boundingRect]).get() ];
```
Cast on iOS is unnecessary I think.

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/WKTextPlaceholder.mm:
```diff
@@ -48,11 +48,19 @@ - (instancetype)initWithElementContext:(const WebCore::ElementContext&)context
-    return @[ adoptNS([[WKTextSelectionRect alloc] initWithCGRect:_elementContext.boundingRect]).get() ];
+#if PLATFORM(IOS_FAMILY)
+    return @[(UITextSelectionRect *)adoptNS([[WKTextSelectionRect alloc] initWithCGRect:_elementContext.boundingRect]).get() ];
+#elif PLATFORM(MAC)
```
Could just be `#else`.

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/WKTextSelectionRect.h:
```diff
@@ -24,18 +24,27 @@
+#if PLATFORM(IOS_FAMILY)
 @interface WKTextSelectionRect : UITextSelectionRect
+#elif PLATFORM(MAC)
+@interface WKTextSelectionRect : NSObject // FIXME: change to NSTextSelectionRect after rdar://126379463 lands
```
Nit: Comment convention.
```suggestion
@interface WKTextSelectionRect : NSObject // FIXME: Change to NSTextSelectionRect after rdar://126379463 lands.
```

pxlcoder commented on Source/WebCore/PAL/pal/spi/mac/NSTextInputContextSPI.h:
```diff
@@ -29,6 +29,24 @@
+
+@interface NSTextPlaceholder : NSObject
+@property (nonatomic, readonly) NSArray<NSTextSelectionRect *> *rects;
+@end
```
I think `NSTextPlaceholder` is already in the SDK, so you could just import the header, and then declare this in a category.

megangardner replied:
While this is true, if it's not going to hurt anything, I'd rather leave it for now, since this is all staging and will be gone soon anyways.

whsieh replied:
I would've thought defining the interface again would result in a duplicate interface declaration 🤔.

pxlcoder commented on Source/WebKit/UIProcess/ios/AppKitSoftLink.h:
```diff
@@ -25,11 +25,15 @@
+#if HAVE(NSTEXTPLACEHOLDER)
+SOFT_LINK_CLASS_FOR_HEADER(WebKit, NSTextPlaceholder)
+SOFT_LINK_CLASS_FOR_HEADER(WebKit, NSTextSelectionRect)
 #endif
```
I think soft linking is unnecessary now that we don't actually allocate a real `NSTextSelectionRect`.

And it looks like `NSTextPlaceholder` is already in the SDK. The soft linking is clearly not needed for that, since you're able to `-[WKTextPlaceholder alloc]` above.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
```diff
@@ -375,6 +373,9 @@ GenerateSyntheticEditingCommand(enum:uint8_t WebKit::SyntheticEditingCommandType
 
+    InsertTextPlaceholder(WebCore::IntSize size) -> (std::optional<WebCore::ElementContext> placeholder)
+    RemoveTextPlaceholder(struct WebCore::ElementContext placeholder) -> ()
+    
```
`#if PLATFORM(COCOA)`

pxlcoder commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -1068,6 +1065,9 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
 #endif
 #endif // PLATFORM(IOS_FAMILY)
 
+    void insertTextPlaceholder(const WebCore::IntSize&, CompletionHandler<void(const std::optional<WebCore::ElementContext>&)>&&);
+    void removeTextPlaceholder(const WebCore::ElementContext&, CompletionHandler<void()>&&);
```
`#if PLATFORM(COCOA)`

pxlcoder commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -12520,6 +12520,24 @@ void WebPageProxy::storeAppHighlight(const WebCore::AppHighlight& highlight)
+    }
+    sendWithAsyncReply(Messages::WebPage::RemoveTextPlaceholder { placeholder }, WTFMove(completionHandler));
+}
+
```
`#if PLATFORM(COCOA)`

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/WKTextSelectionRect.h:
```diff
@@ -24,18 +24,29 @@
  */
 
```
`#if PLATFORM(COCOA)`

whsieh commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4547,6 +4548,27 @@ static NSPasteboardName pasteboardNameForAccessCategory(WebCore::DOMPasteAccessC
+            return;
+        }
+        WebCore::ElementContext placeholderToUse { *placeholder };
+        completionHandler(adoptNS([[WKTextPlaceholder alloc] initWithElementContext:placeholderToUse]).get());
```
Nit - might be a tiny bit simpler as:

```suggestion
        completionHandler(adoptNS([[WKTextPlaceholder alloc] initWithElementContext:*placeholder]).get());
```

whsieh commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4547,6 +4548,27 @@ static NSPasteboardName pasteboardNameForAccessCategory(WebCore::DOMPasteAccessC
+{
+    // FIXME: Implement support for willInsertText. See <https://bugs.webkit.org/show_bug.cgi?id=208747>.
+    if (auto* wkTextPlaceholder = dynamic_objc_cast<WKTextPlaceholder>(placeholder))
+        m_page->removeTextPlaceholder(wkTextPlaceholder.elementContext, makeBlockPtr(completionHandler));
```
Nit - `*` is on the wrong side for an ObjC object. That said, we should probably just deploy `RetainPtr` here:

```suggestion
    if (RetainPtr wkTextPlaceholder = dynamic_objc_cast<WKTextPlaceholder>(placeholder))
        m_page->removeTextPlaceholder([wkTextPlaceholder elementContext], makeBlockPtr(completionHandler));
```

megangardner committed:
Adopt NSTextPlaceholders for Mac.
https://bugs.webkit.org/show_bug.cgi?id=272961
rdar://126696059

Reviewed by Wenson Hsieh.

Adopt NSTextPlaceholders which is a parallel of UITextPlaceholders.
Move the now common code to Cocoa to allow for code sharing.

* Source/WebCore/PAL/pal/spi/mac/NSTextInputContextSPI.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView insertTextPlaceholderWithSize:completionHandler:]):
(-[WKWebView removeTextPlaceholder:willInsertText:completionHandler:]):
* Source/WebKit/UIProcess/Cocoa/WKTextPlaceholder.h: Renamed from Source/WebKit/UIProcess/ios/WKTextPlaceholder.h.
* Source/WebKit/UIProcess/Cocoa/WKTextPlaceholder.mm: Renamed from Source/WebKit/UIProcess/ios/WKTextPlaceholder.mm.
(-[WKTextPlaceholder initWithElementContext:]):
(-[WKTextPlaceholder elementContext]):
(-[WKTextPlaceholder rects]):
* Source/WebKit/UIProcess/Cocoa/WKTextSelectionRect.h: Renamed from Source/WebKit/UIProcess/ios/WKTextSelectionRect.h.
* Source/WebKit/UIProcess/Cocoa/WKTextSelectionRect.mm: Renamed from Source/WebKit/UIProcess/ios/WKTextSelectionRect.mm.
(-[WKTextSelectionRectCustomHandleInfo initWithFloatQuad:isHorizontal:]):
(-[WKTextSelectionRectCustomHandleInfo bottomLeft]):
(-[WKTextSelectionRectCustomHandleInfo topLeft]):
(-[WKTextSelectionRectCustomHandleInfo bottomRight]):
(-[WKTextSelectionRectCustomHandleInfo topRight]):
(-[WKTextSelectionRect initWithCGRect:]):
(-[WKTextSelectionRect initWithSelectionGeometry:scaleFactor:]):
(-[WKTextSelectionRect _path]):
(-[WKTextSelectionRect writingDirection]):
(-[WKTextSelectionRect range]):
(-[WKTextSelectionRect _customHandleInfo]):
(-[WKTextSelectionRect rect]):
(-[WKTextSelectionRect containsStart]):
(-[WKTextSelectionRect containsEnd]):
(-[WKTextSelectionRect isVertical]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::insertTextPlaceholder):
(WebKit::WebPageProxy::removeTextPlaceholder):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/AppKitSoftLink.h:
* Source/WebKit/UIProcess/ios/AppKitSoftLink.mm:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::insertTextPlaceholder): Deleted.
(WebKit::WebPageProxy::removeTextPlaceholder): Deleted.
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::insertTextPlaceholderWithSize):
(WebKit::WebViewImpl::removeTextPlaceholder):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::insertTextPlaceholder):
(WebKit::WebPage::removeTextPlaceholder):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::insertTextPlaceholder): Deleted.
(WebKit::WebPage::removeTextPlaceholder): Deleted.
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextPlaceholderTests.mm: Added.

Canonical link: https://commits.webkit.org/277879@main
